src/app/clusters/mode-select-server/mode-select-delegate.h

```diff
@@ -0,0 +1,113 @@
+/*
+ *
+ *    Copyright (c) 2023 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#ifndef SOME_UNIQUE_NAME_HERE
+#define SOME_UNIQUE_NAME_HERE
+#pragma once
+
+#include <app-common/zap-generated/cluster-objects.h>
+#include <app/CommandHandlerInterface.h>
+#include <app/util/util.h>
+#include <utility>
+#include <vector>
+
+using chip::Protocols::InteractionModel::Status;
+using ModeOptionStructType  = chip::app::Clusters::ModeSelect::Structs::ModeOptionStruct::Type;
+using SemanticTagStructType = chip::app::Clusters::ModeSelect::Structs::SemanticTagStruct::Type;
+
+template <typename T>
+using List = chip::app::DataModel::List<T>;
+
+namespace chip {
+namespace app {
+namespace Clusters {
+namespace ModeSelect {
+
+class Delegate
+{
+public:
+    const std::vector<ModeOptionStructType> modeOptions;
+
+    /**
+     * This is a helper function to build a mode option structure. It takes the label/name of the mode,
+     * the value of the mode and a list of semantic tags that apply to this mode.
+     */
+    static ModeOptionStructType BuildModeOptionStruct(const char * label, uint8_t mode,
+                                                      const List<const SemanticTagStructType> & semanticTags)
+    {
+        Structs::ModeOptionStruct::Type option;
+        option.label        = CharSpan::fromCharString(label);
```

Comment:
So this is assuming that the label has longer lifetime than this delegate, right?   That needs to be at least documented...

In practice, do we want to be storing all this data in RAM instead of having some interface that lets us read it out of storage, or compile-time constants, or whatever other representation the implementation chooses in the end?

Reply:
With the re-implementation of the mode select delegate to remove the use of `std::vector`,  the `BuildModeOptionStruct` is just a helper function that the user can use to help them build their store of mode information. If the user would like not to store the data in RAM, or have some other way they would like to store this information, they can always opt not to use this function and implement their own storage as this is only used by the application specific delegates.

I have added the following note to the documentation to inform the user that they must ensure that the label and semanticTags must stay alive for the entire lifetime of the returned structure. `NOTE, the caller must ensure that the lifetime of the label and semanticTags is as long as the returned structure.`